k2/csrc/host/fsa_util.cc

```diff
@@ -251,46 +262,63 @@ void GetArcIndexes2(const Array2<int32_t *, int32_t> &arc_map,
 
 void StringToFsa::GetSizes(Array2Size<int32_t> *fsa_size) {
   K2_CHECK_NE(fsa_size, nullptr);
-  fsa_size->size1 = fsa_size->size2 = 0;
+  fsa_size->size1 = static_cast<int32_t>(arcs_.size());
+  fsa_size->size2 = num_arcs_;
+}
 
+void StringToFsa::ReadArcsFromString() {
   static constexpr const char *kDelim = " \t";
   std::istringstream is(s_);
   std::string line;
   bool finished = false;  // when the final state is read, set it to true.
-  int32_t num_arcs = 0;
+  num_arcs_ = 0;
   while (std::getline(is, line)) {
     std::vector<std::string> splits;
     SplitStringToVector(line, kDelim, &splits);
     if (splits.empty()) continue;
 
     K2_CHECK_EQ(finished, false);
 
-    auto fields = StringVectorToIntVector(splits);
+    auto fields = StringVectorToFloatVector(splits);
     auto num_fields = fields.size();
-    if (num_fields == 3u) {
+    if (num_fields == 3u || num_fields == 4u) {
       Arc arc{};
-      arc.src_state = fields[0];
-      arc.dest_state = fields[1];
-      arc.label = fields[2];
+      arc.src_state = static_cast<int32_t>(fields[0]);
```

Comment:
We have to check that these are integers, i.e., there are no truncations.

---

https://github.com/k2-fsa/k2/blob/866c6cf9c7440285678e7ca9a51ccc317a7162af/k2/csrc/fsa_utils.cu#L115-L122 may be another alternative.

Reply:
Fixed.